lib/sudo-exec/src/monitor.rs

```diff
@@ -19,82 +19,96 @@ use crate::{
-            backchannel
-                .send(&ParentMessage::CommandPid(command_pid))
-                .ok();
+    // spawn the command.
+    let command = command.spawn()?;
 
-            // Register the callback to receive events from the backchannel
-            dispatcher.set_read_callback(&backchannel, |mc, ev| mc.read_backchannel(ev));
+    let command_pid = command.id() as ProcessId;
 
-            // set the process group ID of the command to the command PID.
-            let command_pgrp = command_pid;
-            setpgid(command_pid, command_pgrp).ok();
+    // Send the command's PID to the parent.
+    backchannel
+        .send(&ParentMessage::CommandPid(command_pid))
+        .ok();
 
-            Ok((dispatcher, command_pid, command_pgrp, command, pty_follower))
-        });
+    let mut closure = MonitorClosure::new(command, command_pid, backchannel, &mut dispatcher);
 
-        match result {
-            Err(err) => {
-                backchannel.send(&err.into()).unwrap();
-                exit(1);
-            }
-            Ok((dispatcher, command_pid, command_pgrp, command, pty_follower)) => (
-                Self {
-                    command_pid: Some(command_pid),
-                    command_pgrp,
-                    command,
-                    _pty_follower: pty_follower,
-                    backchannel,
-                },
-                dispatcher,
-            ),
-        }
-    }
+    // FIXME (ogsudo): Here's where the signal mask is removed because the handlers for the signals
+    // have been setup after initializing the closure.
+    // FIXME (ogsudo): Set the command as the foreground process for the follower.
+
+    // Start the event loop.
+    dispatcher.event_loop(&mut closure);
+    // FIXME (ogsudo): Terminate the command using `killpg` if it's not terminated.
+    // FIXME (ogsudo): Take the controlling tty so the command's children don't receive SIGHUP when we exit.
+    // FIXME (ogsudo): Send the command status back to the parent.
+    // FIXME (ogsudo): The tty is restored here if selinux is available.
 
-    pub(super) fn run(mut self, dispatcher: &mut EventDispatcher<Self>) -> ! {
-        dispatcher.event_loop(&mut self);
-        drop(self);
-        exit(0);
+    drop(closure);
+
+    exit(1)
+}
+
+struct MonitorClosure<'a> {
+    /// The command PID.
+    ///
+    /// This is `Some` iff the process is still running.
+    command: Child,
+    command_pid: Option<ProcessId>,
+    command_pgrp: ProcessId,
+    backchannel: &'a mut MonitorBackchannel,
+}
+
+impl<'a> MonitorClosure<'a> {
+    fn new(
+        command: Child,
+        command_pid: ProcessId,
+        backchannel: &'a mut MonitorBackchannel,
+        dispatcher: &mut EventDispatcher<Self>,
+    ) -> Self {
+        // FIXME (ogsudo): Store the pgid of the monitor.
+
+        // Register the callback to receive events from the backchannel
+        dispatcher.set_read_callback(backchannel, |mc, ev| mc.read_backchannel(ev));
```

Comment:
the closure could have more readable variable names

Reply:
huh, how did this survive the other PR? I'll fix it